### PR TITLE
HashingSource and HashingSink.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: okio/src/main/java/okio/HashingSink.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  /** Returns a sink that uses the SHA-256 hash algorithm. */
+  public static HashingSink sha256(Sink sink) {
+    try {
+      return new HashingSink(sink, MessageDigest.getInstance("SHA-256"));
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  private HashingSink(Sink sink, MessageDigest messageDigest) {
```
Comment (JakeWharton) [inline okio/src/main/java/okio/HashingSink.java]: No factories for supplying your own `MessageDigest`?

File: okio/src/main/java/okio/HashingSource.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  /** Returns a sink that uses the obsolete SHA-1 hash algorithm. */
+  public static HashingSource sha1(Source source) {
+    try {
+      return new HashingSource(source, MessageDigest.getInstance("SHA-1"));
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  /** Returns a sink that uses the obsolete MD5 hash algorithm. */
```
Comment (JakeWharton) [inline okio/src/main/java/okio/HashingSource.java]: These are out in reverse order compared to the other one :no_mouth:

**Comment by JakeWharton** [comment]
Looks great! Very close to hypothetical `Sha1Sink` from my talk a year ago except with nicer types. :shipit:

File: okio/src/main/java/okio/HashingSource.java (hunk 2 above)
Comment (swankjesse) [inline okio/src/main/java/okio/HashingSource.java]: Fixed. A-Z for now.

File: okio/src/main/java/okio/HashingSink.java (hunk 1 above)
Comment (swankjesse) [inline okio/src/main/java/okio/HashingSink.java]: None. In theory we later add hashes not-exposed by MessageDigest like MurmurHash3, so I don’t want to couple this to that API.

**Comment by swankjesse** [comment]
Build failure is flakiness in attempting to build a shaded jar for some JMH deps. Sigh.

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:2.2:shade (default) on project benchmarks: Error creating shaded jar: error in opening zip file /home/travis/.m2/repository/org/apache/commons/commons-math3/3.2/commons-math3-3.2.jar -> [Help 1]
```

**Pull request merged**